File: web/src/api/tax.ts
import { apiFetch, withEntityId } from './client'

export interface TaxCalculateRequest {
  entity_id?: string
  start_date: string
  end_date: string
}

export interface TaxSummaryResponse {
  period_start: string
  period_end: string
  total_realized_gain_usd: number
  total_transfer_tax_vnd: number
  total_exempt_vnd: number
  closed_lot_count: number
  open_lot_count: number
  taxable_transfer_count: number
}

export interface ClosedLotResponse {
  symbol: string
  quantity: number
  cost_basis_usd: number
  proceeds_usd: number
  gain_usd: number
  holding_days: number
  buy_date: string
  sell_date: string
}

export interface OpenLotResponse {
  symbol: string
  remaining_quantity: number
  cost_basis_per_unit_usd: number
  buy_date: string
}

export interface TaxableTransferResponse {
  timestamp: string
  symbol: string
  quantity: number
  value_vnd: number
  tax_amount_vnd: number
  exemption_reason: string | null
}

export interface TaxCalculateResponse {
  summary: TaxSummaryResponse
  closed_lots: ClosedLotResponse[]
  open_lots: OpenLotResponse[]
  taxable_transfers: TaxableTransferResponse[]
}

export function calculateTax(body: TaxCalculateRequest) {
  return apiFetch<TaxCalculateResponse>('/tax/calculate', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export interface RealizedGainsFilters {
  symbol?: string
  date_from?: string
  date_to?: string
  gain_only?: boolean
  loss_only?: boolean
}

export interface OpenLotsFilters {
  symbol?: string
  min_quantity?: number
}

export function getRealizedGains(entityId?: string, filters: RealizedGainsFilters = {}) {
  const params = new URLSearchParams()
  if (filters.symbol) params.set('symbol', filters.symbol)
  if (filters.date_from) params.set('date_from', filters.date_from)
  if (filters.date_to) params.set('date_to', filters.date_to)
  if (filters.gain_only !== undefined) params.set('gain_only', String(filters.gain_only))
  if (filters.loss_only !== undefined) params.set('loss_only', String(filters.loss_only))
  const qs = params.toString()
  const path = `/tax/realized-gains${qs ? `?${qs}` : ''}`
  return apiFetch<ClosedLotResponse[]>(withEntityId(path, entityId))
}

export function getOpenLots(entityId?: string, filters: OpenLotsFilters = {}) {
  const params = new URLSearchParams()
  if (filters.symbol) params.set('symbol', filters.symbol)
  if (filters.min_quantity !== undefined) params.set('min_quantity', String(filters.min_quantity))
  const qs = params.toString()
  const path = `/tax/open-lots${qs ? `?${qs}` : ''}`
  return apiFetch<OpenLotResponse[]>(withEntityId(path, entityId))
}

export function getTaxSummary(entityId?: string) {
  return apiFetch<TaxSummaryResponse | null>(withEntityId('/tax/summary', entityId))
}

File: src/cryptotax/api/schemas/tax.py
"""Pydantic schemas for tax API."""

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


class TaxCalculateRequest(BaseModel):
    entity_id: uuid.UUID | None = None  # None = use default entity
    start_date: str  # ISO format: "2025-01-01"
    end_date: str  # ISO format: "2025-12-31"


class ClosedLotResponse(BaseModel):
    symbol: str
    quantity: Decimal
    cost_basis_usd: Decimal
    proceeds_usd: Decimal
    gain_usd: Decimal
    holding_days: int
    buy_date: datetime
    sell_date: datetime


class OpenLotResponse(BaseModel):
    symbol: str
    remaining_quantity: Decimal
    cost_basis_per_unit_usd: Decimal
    buy_date: datetime


class TaxableTransferResponse(BaseModel):
    timestamp: datetime
    symbol: str
    quantity: Decimal
    value_vnd: Decimal
    tax_amount_vnd: Decimal
    exemption_reason: str | None = None


class TaxSummaryResponse(BaseModel):
    period_start: datetime
    period_end: datetime
    total_realized_gain_usd: Decimal
    total_transfer_tax_vnd: Decimal
    total_exempt_vnd: Decimal
    closed_lot_count: int
    open_lot_count: int
    taxable_transfer_count: int


class TaxCalculateResponse(BaseModel):
    summary: TaxSummaryResponse
    closed_lots: list[ClosedLotResponse]
    open_lots: list[OpenLotResponse]
    taxable_transfers: list[TaxableTransferResponse]
